### DISCUSSION TITLE
Scala-cli REPL with Toolkit

### DISCUSSION THREAD

**Discussion by spamegg1** [scala/toolkit#37] [Q&A] [answered]
Is it possible to fire up a Scala-cli REPL with the Toolkit loaded?
```bash
scala-cli repl --toolkit                                                                                          
Argument --toolkit: argument missing

To list all available options, run
  scala-cli repl --help
```
I'm guessing, not? Could not find anything in the documentation either.

**Comment by spamegg1** [comment, marked as answer]
Never mind, looks like it IS possible:
```bash
╰─ scala-cli repl --toolkit=latest                                                                                   
Downloading 3 dependencies and 2 internal dependencies
Welcome to Scala 3.3.1 (17.0.7, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                        
scala> 
```
It would be nice if this was documented somewhere. It's not [here](https://scala-cli.virtuslab.org/docs/cookbooks/) or [here](https://docs.scala-lang.org/toolkit/introduction.html). 

I will leave this here in case anyone else wonders.

**Comment by som-snytt** [comment]
I'm still not a user, only an onlooker, but it seems I have some version installed.

> Pass --help to any sub-command of Scala CLI to list its options:

https://scala-cli.virtuslab.org/docs/guides/configuration#command-line

```
➜  scala-cli repl --help
Usage: scala-cli repl [options]
Fire-up a Scala REPL.

You are currently viewing the basic help for the repl sub-command. You can view the full help by running:
   scala-cli repl --help-full
For detailed documentation refer to our website: https://scala-cli.virtuslab.org/docs/commands/repl

Scala options:
  -P, --plugin, --compiler-plugin string*        Add compiler plugin dependencies
  -S, --scala, --scala-version version           Set the Scala version (3.2.2 by default)
  --platform scala-js|scala-native|jvm           Specify platform
  -with-compiler, --with-compiler, --with-scala-compiler  Allows to include the Scala compiler artifacts on the classpath.
  -d, --compile-out, --destination, --compile-output, --output-directory, --compilation-output /example/path  Copy compilation results to output directory using either relative or absolute path
  --toolkit, --with-toolkit version|latest       Add toolkit to classPath
```

**Reply by spamegg1** [reply]
Thanks, I actually looked at `scala-cli repl --help` but I didn't see `--toolkit` all the way at the bottom there! It's so well hidden! :secret: 

I think this should definitely be much, much, much more visible, especially in the [REPL section](https://scala-cli.virtuslab.org/docs/commands/repl) and in the [Cookbook](https://scala-cli.virtuslab.org/docs/cookbooks/intro).

It's a bit weird, since [Toolkit docs](https://docs.scala-lang.org/toolkit/introduction.html#how-to-run-the-code) don't show how to use *any one of the 6 methods mentioned*, and [Scala-cli docs](https://scala-cli.virtuslab.org/docs/) don't focus on the Toolkit. It's extremely confusing for a newcomer.

This does not explain, therefore assumes, some "default / obvious" way of loading the Toolkit into the REPL:
![Screenshot from 2023-11-09 22-13-10](https://github.com/scala/toolkit/assets/4255997/97eea0ed-1016-46b2-a82b-79fca658da17)
This mentions scripts with Scala-cli, not the REPL:
![Screenshot from 2023-11-09 22-14-31](https://github.com/scala/toolkit/assets/4255997/6bb0a741-5670-4c30-9a05-a5c08ba2d777)

So, is there a non-Scala-cli, non-script way to load the Toolkit into the REPL? I'm not aware of it. That's why Scala-cli and the Toolkit are so closely linked in my mind. (I was under the impression they are made by the same people.)

**Reply by som-snytt** [reply]
For me (as a very casual user), searching around the docs is a cross between git and sbt UX.

They should sell "doesn't save your code anywhere" as a privacy feature that saves you embarrassment later.

On CLI help, a big reason for `-V -W -X -Y` in scalac was to reduce scanning lists of options. In scala 3, I want `-Vphases` to take a phase name (or partial name) and highlight it in the huge list, to save grepping in output, which might be on stderr.

**Reply by spamegg1** [reply]
> They should sell "doesn't save your code anywhere" as a privacy feature that saves you embarrassment later.

:laughing: :+1: 

> For me (as a very casual user), searching around the docs is a cross between git and sbt UX.

Just adding a one-line `scala-cli repl --toolkit=latest` [here](https://docs.scala-lang.org/toolkit/introduction.html#how-to-run-the-code) would go ***a long way.*** (This is an instance of a more general "agnosticism in docs" issue that is seen everywhere in tech.)  I spent near a day trying to make that (or the third option, worksheets) work.

**Reply by SethTisue** [reply]
> is there a non-Scala-cli, non-script way to load the Toolkit into the REPL

The Toolkit is just an ordinary dependency. It transitively depends on the included libraries.

So in a build tool such as sbt, if add the toolkit to your `libraryDependencies`, then `sbt console` will start a REPL where the toolkit is available.

I think Mill also lets you start a Scala REPL with your code and dependencies on the classpath. Gradle and Maven, I'm not sure about.

**Reply by SethTisue** [reply]
> [Toolkit docs](https://docs.scala-lang.org/toolkit/introduction.html#how-to-run-the-code) don't show how to use any one of the 6 methods mentioned

That's not entirely true, but I think it's very understandable that you didn't see it.

At the top of every "How do I....?" page in the Toolkit docs is a twistdown like this:

> <img width="781" alt="Screenshot 2023-11-09 at 2 02 27 PM" src="https://github.com/scala/toolkit/assets/161079/9530dd2c-d194-4244-8d28-5ef48a8bd154">

And if you open the twistdown, you see instructions with tabs for scala-cli, sbt, and mill:

> <img width="789" alt="Screenshot 2023-11-09 at 2 03 18 PM" src="https://github.com/scala/toolkit/assets/161079/552e0cc8-d966-45e5-b613-8da48b95426e">


so it's there, but if you missed it, then apparently the UX of how it's presented isn't ideal? fyi @szymon-rd @adpi2

**Reply by som-snytt** [reply]
I tried those widgets, and before I forget again, I intended to suggest that if there is a correspondence between "in-source comment" and "command line", then that could be summarized. An issue with `using options` is what's the semantics for `foo` where foo is `toolkit`. So the widgets don't cover when I'm just running repl from cli. I'm in late afternoon fog, so maybe I forgot what I saw earlier.

As always, worth adding, great effort on docs, I totally get FOMO. Just finding this one thing is always a hard ask.

**Reply by spamegg1** [reply]
That's fair @SethTisue I should have looked more carefully. And yes, the "Getting" tab is a bit confusing.

I still don't think I would have figured out `scala-cli repl --toolkit=latest` easily. I actually got it by chance after a lot of trial and error. It wasn't in Toolkit docs or Scala-cli docs, yes I should have looked more carefully into the command line options, and I couldn't think of the possibility to combine a sub-command (`repl`) with a command line option (`--toolkit=latest`) at first... 

When the [docs](https://docs.scala-lang.org/toolkit/introduction.html) say "interactively in the Scala REPL" I used the `scala` command (that's what REPL means to me), and I thought Toolkit would be already loaded into the REPL.

```scala
 ➜ scala
Welcome to Scala 3.3.1 (17.0.7, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                      
scala> import os.*
-- [E006] Not Found Error: -------------------------------------------------------------------------------------------
1 |import os.*
  |       ^^
  |       Not found: os
  |
  | longer explanation available when compiling with `-explain`
1 error found
```

I know Scala-cli became the official runner but still, `scala` (the default REPL) and `scala-cli` are separate commands, right? Or are they aliases for each other now? Maybe there is something wrong with my setup then? I got Scala 3.3.1 from SDKMAN (in fact I was a bit involved [in its release](https://github.com/lampepfl/dotty/issues/18515#issuecomment-1722845980)) I guess SDKMAN does not install Scala-cli along with the language?

**Comment by som-snytt** [comment]
Please reconcile with https://github.com/VirtusLab/scala-cli/issues/2426 deprecate latest.

**Comment by spamegg1** [comment]
I think it's my misunderstanding that Scala-cli became the default code runner, apparently that's not going to happen until 3.4 is released. So the docs make a bit more sense now (they are future-leaning).

**Comment by som-snytt** [comment]
In 2026, I found this discussion again, thanks.

```
scala-cli repl --server=false --with-toolkit -S 3.8.3 -Vprint:typer
[error]  3.8.3: input file not found
Try viewing the relevant help to see the list of available sub-commands and options.
  scala-cli repl --help
```

I just wanted to try a quick test for `scala-xml`, which of course it turns out the toolkit does not support after all.
